pkg/localprovider/localprovider.go

```diff
@@ -0,0 +1,80 @@
+package localprovider
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	upgradev1alpha1 "github.com/openshift/managed-upgrade-operator/pkg/apis/upgrade/v1alpha1"
+	"github.com/openshift/managed-upgrade-operator/util"
+)
+
+const UPGRADECONFIG_CR_NAME = "managed-upgrade-config"
+
+var log = logf.Log.WithName("upgradeconfig-localprovider")
+
+func New(c client.Client, name string) (*localProvider, error) {
+	return &localProvider{
+		client:  c,
+		cfgname: name,
+	}, nil
+}
+
+type localProvider struct {
+	client  client.Client
+	cfgname string
+}
+
+// Get checks the upgrade config on the cluster with matched name which is going to perform the upgrade
+func (l *localProvider) Get() ([]upgradev1alpha1.UpgradeConfigSpec, error) {
+	log.Info("Read the upgrade config from the cluster directly")
+
+	// Get the current UpgradeConfigs on the cluster
+	instances, err := fetchUpgradeConfigs(l.client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get Specs from the upgradeConfig
+	specs, err := readSpecFromConfig(*instances)
+	if err != nil {
+		return nil, err
+	}
+
+	return specs, nil
+}
+
+// Helper function to extract the spec from the upgradeConfig CR
+func readSpecFromConfig(ucl upgradev1alpha1.UpgradeConfigList) ([]upgradev1alpha1.UpgradeConfigSpec, error) {
+	upgradeConfigSpecs := make([]upgradev1alpha1.UpgradeConfigSpec, 0)
+
+	for _, u := range ucl.Items {
+		// Completed UpgradeConfigs can be ignored
+		history := u.Status.History.GetHistory(u.Spec.Desired.Version)
+		if history.Phase != upgradev1alpha1.UpgradePhaseUpgraded {
```

Comment:
I just thought of one more thing that I forgot, sorry. :joy: 
To be panic safe it should maybe check that the history returned isn't nil first.

```suggestion
		if history != nil && history.Phase != upgradev1alpha1.UpgradePhaseUpgraded {
```